### ISSUE TITLE
Convert hooks API to use a monadic style (State Monad)

### ISSUE THREAD

**Issue by entropitor** [revery-ui/revery#564]
If we were to convert hooks into a monadic style ([state monad](https://wiki.haskell.org/State_Monad)), we will be able to start using `let+` and `let*` when they land in Reason and we would not need to keep passing hooks around manually

```reason
let componentWithHooks = {
    let component = React.component("componentWithHooks");
    (~children, ()) => component(Hooks.{
       let* (state, setState) = state'(0);
       <Button text=string_of_int(state) onClick={(_) => setState(state + 1)}/>
    });
};
```

A new version of state' should look something like this:
```reason
let state' = (initialState, hooksState) => {
  let* hooks = get hooksState;
  let (state, setState, hooks) = Hooks.state(initialState, hooks);
  put hooks;
  return (state, setState)
}
```

And the component function should extract the hooks object from the state and the component from the return value

**Comment by bryphe** [comment]
Thanks for the suggestion, @entropitor ! @wokalski and @let-def might be interested too - this was something we discussed briefly at the Revery workshop.

How would this composition work with other hook types, like `effect`, `ref`, and `reducer`?

@wokalski is also investigating a PPX to streamline this syntax

**Comment by entropitor** [comment]
I thought about it a bit more and using a state monad exactly might not work entirely because of the GADT.

But you basically have this list of hooks (https://github.com/briskml/brisk-reconciler/blob/master/lib/Hooks.re#L3-L6) which is your state. Let me illustrate with a random number generator (RNG). In a purely functional language like haskell, this would need to use a state monad because of similar reasons (https://en.wikibooks.org/wiki/Haskell/Understanding_monads/State)

So getting 2 random numbers could look like something like this:
```reason
let g = RandomNumberGenerator.make();
let (number1, g1) = g.nextInt()
let (number2, g2) = g1.nextInt()
```

With state monad and new `let+` binding in ocaml, it would look more like
```
open RandomNumberGenerator.LetPlusSyntax;
RandomNumberGenerator.make();
let+ number1 = nextInt()
let+ number2 = nextInt()
```

So basically `effect`, `ref`, and `reducer` should all have very similar implementations and the real magic goes on in the let+ binding. So the user might need to open a module to include the right let+ binding but that should be everything. And then we wouldn't need any ppx (temporarily we could have the https://github.com/jaredly/let-anything, until reason fully supports let+ syntax)

**Comment by wokalski** [comment]
@entropitor great suggestion, thanks! One thing to remember is that at no point do we have "Hooks.make". We get hooks as a function parameter and it is a constraint. We can resolve it by using `return` at the end instead of `make` at the beginning like so:

```reason
let%hook (state, setState) = Hooks.state(false);
let%hook () = Hooks.effect(Always, doEffect);
return(
  if (state) {
    <Div> <Text title="cell1" /> <Text title="cell2" /> </Div>;
  } else {
    <Div> <Text title="well" /> </Div>;
  }
)
```

```reason
  let bind = (callHook, ~f) => hooks => {
    let (value_, nextHooks) = callHook(hooks);
    f(value_, nextHooks);
  };

  let return = (x, hooks) => (hooks, x);

  let component = () => {
    bind(state'(false), ~f=((state, setState)) => {
      bind(
        effect'(Always, doEffect),
        ~f=() =>
        return(
          if (state) {
            <Div> <Text title="cell1" /> <Text title="cell2" /> </Div>;
          } else {
            <Div> <Text title="well" /> </Div>;
          },
        )
      )
    });
  };
```

**Comment by glennsl** [comment]
Isn't `React.component` essentially `Hooks.make`?

**Comment by wokalski** [comment]
What do you mean @glennsl ?

**Comment by glennsl** [comment]
Today you create a "component" with `React.component`. You then pass the component a function, which receives `hooks` as a parameter, allowing you to create hooks within the context of that function. Something like this:

```
let component = React.component("componentWithHooks");
let make = (~children, ()) =>
  component(hooks => ...);
```

So isn't `React.component` then basically `Hooks.make`?

**Comment by entropitor** [comment]
I think https://github.com/revery-ui/revery/issues/564#issuecomment-537215122 is more or less what I meant yes!

And @glennsl, I agree, potentially we can make a convenient API around it, but I think the most important goal is to nail the inner API first and then we can potentially create a new top-level API to better match this (if necessary).

**Comment by wokalski** [comment]
@entropitor @glennsl Your thought with using component as make is sort of intriguing. What I'm trying to focus on though, is eliminating the `return` in favour of wrapping it all before the first hook call. I can't figure it out though...

I'd like to have something  like this:

```
  let component = () => {
    return (
    not_bind(state'(false), ~f=((state, setState)) => {
      not_bind(
        effect'(Always, doEffect),
        ~f=() =>
          if (state) {
            <Div> <Text title="cell1" /> <Text title="cell2" /> </Div>;
          } else {
            <Div> <Text title="well" /> </Div>;
          }
      )
    }));
  };
```

The rationale is simple - it'll be much easier to develop a PPX which wraps the first hook call with something extra than one that finds the last hooks call.

**Comment by glennsl** [comment]
Right, I was mostly thinking it would be easier for people to transition with `React.component` still in place instead of `return`, but wasn't considering any kind of ppx, which might make that moot anyway.

Carry on then! :)

**Comment by entropitor** [comment]
```reason
// React.re
let componentWithHooks = (displayName, makeFn) => {
  let component = React.component(displayName)
  component(hooks => {
    // Possibly inject hooks as some sort of hidden state monad into the makeFn as param
    let newHooksThingy = makeNewHooksThingy(hooks)
    let result = makeFn(newHooksThingy);
    // Possibly extract hooks as some sort of hidden state out of the monad (out of the result value)
    (getJsxFromResult(result), getHooksFromResult(result))
  })
}
```

The cool thing is that even though hooks is a passed to the component, the user doesn't need to handle it, the bind you wrote will use it to capture it internally in the "hidden" state.

(I'm 99% sure that we can write this with the let_anything ppx and no additional ppx)

I'm not even sure if you need `componentWithHooks` it all depends on whether you need to wrap the hooks object into something else or not. From the bind you wrote earlier, I've the feeling we might not even need it.

**Comment by wokalski** [comment]
That's correct, the bind function I wrote handles this for us. However, as I said, I want to get rid of return. I'd be curious to hear any ideas on that. Also, we're going to provide our own PPX in order to have clear `let%hook` instance and to allow future modifications if needed without changing user API.

**Comment by wokalski** [comment]
@glennsl If you have any ideas how to, err, "invert it" (I'm lacking the vocabulary here) so that call to component is used instead of return, I'd be curious to hear it. If call to `component` happens before any hooks are used, and there's no `return` it'll be very easy to create a ppx for that.

**Comment by bryphe** [comment]
Was talking to @zindel at ReasonConf about this - he had ideas around this too-

**Comment by wokalski** [comment]
Resolved in https://github.com/briskml/brisk-reconciler/pull/42

**Comment by bryphe** [comment]
The new PPX by @wokalski looks amazing! 😍 

![image](https://user-images.githubusercontent.com/13532591/67152266-34b08080-f287-11e9-827a-7f0d8876d4a2.png)
